### ISSUE TITLE
Missing Csound Opcodes - scans, scanu, xscans, xscanu, xscanmap, xscansmap, STKopcodes (bowed, clarinet, brass, etc.)

### ISSUE THREAD

**Issue by csounder** [BelaPlatform/Bela#635]
As my students and I are running instruments we developed and instruments from The Csound Manual on the Bela, we have discovered that some Csound Opcodes, are not included in the version of Csound that is running on the Bela 6.14

- for example, none of the STK opcodes are there (and these are super cool)
- for example, none of the sc_ opcodes are there (and these are ports of SuperCollider opcodes that are very cool)
- most importantly, the main scanned synthesis opcodes are not there - and they are a part of my teaching every semester
(scans, scanu, xscans, xscanu, xscanmap, xscansmap) [we are happy to find scanhammer and scantable, but these are less versatile.)

**Comment by giuliomoro** [comment]
It would seem that there is need for a rethinking of the build system of Csound for Bela. From the [CMake files](https://github.com/csound/csound/blob/develop/Bela/Custom.cmake.bela), there doesn't seem to be any opcode specifically disabled for the Bela build, so it must just be a matter of what libraries are or not installed. 

`grep`'ing through the `csound` source, it would seem that the `sc_` ugens should be in `libscugens.so`, which is indeed on the board at `/usr/lib/csound/plugins-6.0/libscugens.so`. Also, the "scanned synthesis" opcodes, seem to be in `libscansyn.so`, which is on the board at `/usr/lib/csound/plugins-6.0/libscansyn.so`.

So yes, probably it's just a matter of this path not being loaded. Try running this to fix it:

```
ln -s /usr/lib/csound/ /usr/local/lib/csound
```

it gets rid of the warning for me.

**Comment by giuliomoro** [comment]
the STK ones are not there though, I guess that's because they rely on STK which is not installed on Bela (or on the build machine)

**Comment by csounder** [comment]
YES!  This works - and a lot of "other" manual examples and opcodes will not "run" on the Bela.  THANKS.
- it would be cool if the next build of Csound could also include a build of the STK opcodes too.  They sound pretty great and have a lot of rich and interesting "control" possibilities - very mutable and morphable.

**Comment by csounder** [comment]
Two new warnings appear now - which I am sure that you saw - pulseaudio and image opcode libraries are not found (or not compiled?)

WARNING: [mcould not open library '/usr/local/lib/csound/plugins-6.0/libimage.so' (libpng16.so.16: cannot open shared object file: No such file or directory)
WARNING: could not open library '/usr/local/lib/csound/plugins-6.0/librtpulse.so' (libpulse.so.0: cannot open shared object file: No such file or directory)[m

**Comment by giuliomoro** [comment]
right, not sure those are relevant though: `librtpulse` seems about communicating with `pulseaudio`, which is not available on Bela.
`libimage` seems to be about manipulating images ... not sure how that would be useful on Bela.

So if I understand correctly above, I'd suggest we just omit these libraries from the next release.

**Comment by csounder** [comment]
The image opcodes scan and map color data from images to produce control
voltages - they might actually be useful on the Bela.
- maybe you could see the image in the IDE (a jpeg or gif or..) and then
csound instruments would scan it and map from it on the Bela when uploaded
there.


*Dr. Richard Boulanger*

*Professor of Electronic Production and Design*

Professional Writing and Technology Division

*Berklee College of Music*

**Comment by csounder** [comment]
Just omitting the PulseAudio library makes sense since it is not supported
on Bela.

*Dr. Richard Boulanger*

**Comment by giuliomoro** [comment]
> The image opcodes scan and map color data from images to produce control
voltages - they might actually be useful on the Bela.

OK, then it would seem that - assuming your board is connected to the internet - you could do:
```
 apt-get install libpng16-16
```
if the board is not connected to the internet, you'll have to download [this file](http://deb.debian.org/debian/pool/main/libp/libpng1.6/libpng16-16_1.6.28-1+deb9u1_armhf.deb), move it to the board and do `dpkg -i libpng16-16_1.6.28-1+deb9u1_armhf.deb` to install it manually.

> Just omitting the PulseAudio library makes sense since it is not supported

OK. To get rid of the warning, just delete the file for now:

```
 rm /usr/local/lib/csound/plugins-6.0/librtpulse.so
```

**Comment by csounder** [comment]
OK.  librtpulse is gone.  But.  I don't think that Bela is connected to the internet, and so, I will try that option now. THANKS.
- getting there and a lot of new things here to share with the students in our Bela workgroup meeting tomorrow!!!

**Comment by csounder** [comment]
when you recommend that i move the file to the board, should I put it in the /usr/local/lib/csound/plugins-6.0 directory before running the dpkg command?

**Comment by giuliomoro** [comment]
no, it can be anywhere. It's a `.deb` package that will be installed with the `dpkg` command (this in turn will copy its contents to the appropriate paths on the file system). After installation you may well delete the `.deb` file. Oh, also when running `dpkg -i` you have to provide the correct path to wherever the `.deb` is.

**Comment by csounder** [comment]
installed and deleted.  

great!

**Comment by csounder** [comment]
Success - Image Opcodes and Image Opcode Example .csd files from The Csound Manual are running on the Bela
- AWESOME.  And you can see the .png files displayed in the Bela IDE too!

**Comment by csounder** [comment]
would it be this easy to add the STK libraries too?  That would be nice if they were there.  But... I am so happy with all the additions from today.  WOW.  My students will be thrilled tomorrow.  THANKS AGAIN.  -dB

**Comment by giuliomoro** [comment]
Unfortunately not. In the `image` opcodes case, the plugin had been built, but it was linking to a missing library. The step above installed that library. The STK plugin has not been built, therefore this requires re-building csound.

**Comment by csounder** [comment]
Hopefully on the next release.  Thanks again for everything today.  So many new possibilities now!

Dr. Richard Boulanger
Professor
Electronic Production and Design
Berklee College of Music

**Comment by giuliomoro** [comment]
thanks for raising the issues. As we do not use Csound ourselves, issue reporting by users like you is the only way for us to improve.
